ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs

```diff
@@ -434,6 +448,187 @@ traceLedgerProcessing
+
+        currentRtsStats <- GC.getRTSStats
+        let
+          slotDataPoint = DP.mkSlotDataPoint
+                            (realPointSlot rp)
+                            (slot `slotCount` getTipSlot prevLedgerState)
+                            (GC.elapsed_ns         currentRtsStats `nsDiff` GC.elapsed_ns prevRtsStats)
+                            (GC.mutator_elapsed_ns currentRtsStats `nsDiff` GC.mutator_elapsed_ns prevRtsStats)
+                            (GC.gc_elapsed_ns      currentRtsStats `nsDiff` GC.gc_elapsed_ns prevRtsStats)
+                            (GC.major_gcs          currentRtsStats    -     GC.major_gcs prevRtsStats)
+                            (tForecast `div` 1000)
+                            (tHdrTick  `div` 1000)
+                            (tHdrApp   `div` 1000)
+                            (tBlkTick  `div` 1000)
+                            (tBlkApp   `div` 1000)
+          -- Compute the difference between two 'RtsTime's and convert it to microseconds.
+          nsDiff t t' = (t - t') `div` 1000
+
+          slotCount (SlotNo i) = \case
+            Slotting.Origin        -> i
+            Slotting.At (SlotNo j) -> i - j
+
+        IO.hPutStrLn outFileHandle (   DP.showData slotDataPoint separator
+                                    ++ intercalate separator (HasAnalysis.blockStats blk)
+                                   )
+
+        pure $ BenchmarkLedgerOpsState currentRtsStats $ ExtLedgerState ldgrSt' hdrSt'
+      where
+        rp = blockRealPoint blk
+
+        forecast ::
+             SlotNo
+          -> ExtLedgerState blk
+          -> IO (Ticked (LedgerView (BlockProtocol blk)), GC.RtsTime)
+        forecast slot st = time chrono $ do
+          let forecaster = ledgerViewForecastAt lcfg (ledgerState st)
+          case runExcept $ forecastFor forecaster slot of
+            Left err -> fail $ "benchmark doesn't support headers beyond the forecast limit: " <> show rp <> " " <> show err
+            Right !x -> pure x
+
+        tickTheHeaderState ::
+             SlotNo
+          -> ExtLedgerState blk
+          -> Ticked (LedgerView (BlockProtocol blk))
+          -> IO (Ticked (HeaderState blk), GC.RtsTime)
+        tickTheHeaderState slot st tickedLedgerView = time chrono $
+          pure $! tickHeaderState ccfg
+                                  tickedLedgerView
+                                  slot
+                                  (headerState st)
+
+        applyTheHeader ::
+             Ticked (LedgerView (BlockProtocol blk))
+          -> Ticked (HeaderState blk)
+          -> IO (HeaderState blk, GC.RtsTime)
+        applyTheHeader tickedLedgerView tickedHeaderState = time chrono $ do
+          case runExcept $ validateHeader cfg tickedLedgerView (getHeader blk) tickedHeaderState of
+            Left err -> fail $ "benchmark doesn't support invalid headers: " <> show rp <> " " <> show err
+            Right x -> pure x
+
+        tickTheLedgerState ::
```

Comment:
I personally find the local bindings a lot more noisy and distracting than just inlining them into the above.

I feel that most strongly about the two `tick*` bindings: they'd be very legible one-liners if inlined.

Reply:
I introduced the local bindings so that it becomes clear how the data flows between the different steps. If I inline those then we will lose this. I think this gives the reader an idea of the main steps involved, and she can go to the local definitions for more details:

```haskell
        (!tkLdgrView, tForecast) <- forecast            slot prevLedgerState
        (!tkHdrSt,    tHdrTick)  <- tickTheHeaderState  slot prevLedgerState tkLdgrView
        (!hdrSt',     tHdrApp)   <- applyTheHeader                           tkLdgrView tkHdrSt
        (!tkLdgrSt,   tBlkTick)  <- tickTheLedgerState  slot prevLedgerState
        (!ldgrSt',    tBlkApp)   <- applyTheBlock                                       tkLdgrSt
```

Reply:
In particular, after doing this refactoring I realized that for ticking the ledger state and applying the block we didn't need the other steps.